packages/rich-list-core/__tests__/RichListCore.test.ts

```diff
@@ -0,0 +1,111 @@
+    await container.stop()
+  })
+
+  describe('startOrResume() - should start crawl blocks and push active addresses', () => {
+    it('should extract all addresses from transactions and push into queue', async () => {
+      richListCore.start()
+      await waitForCatchingUp(richListCore)
+
+      const queue = await richListCore.queueClient.createQueueIfNotExist('RichListCore_ACTIVE_ADDRESSES', 'LIFO')
+      const queuedItem = await queue.receive(Number.MAX_SAFE_INTEGER)
+      expect(queuedItem).toStrictEqual(EXPECTED_RICH_LIST_ADDRESSES)
+    })
+  })
+
+  describe('calculateNext() + get()', () => {
+    it('should process queued up addresses and update rich list', async () => {
+      richListCore.start()
+      await waitForCatchingUp(richListCore)
+
+      const beforeProcess = await richListCore.get('-1')
+      expect(beforeProcess.length).toStrictEqual(0)
+
+      await richListCore.calculateNext()
+
+      const queue = await richListCore.queueClient.createQueueIfNotExist('RichListCore_ACTIVE_ADDRESSES', 'LIFO')
+      const anythingRemainedInQ = await queue.receive(Number.MAX_SAFE_INTEGER)
+      expect(anythingRemainedInQ.length).toStrictEqual(0)
+
+      const richList = await richListCore.get('-1')
+      expect(richList.length).toStrictEqual(EXPECTED_RICH_LIST_ADDRESSES.length)
+
+      for (let i = 0; i < richList.length - 1; i++) {
+        const current = richList[i]
+        const next = richList[i + 1]
+        expect(EXPECTED_RICH_LIST_ADDRESSES).toContain(current.address)
+        expect(EXPECTED_RICH_LIST_ADDRESSES).toContain(next.address)
+        expect(current.amount).toBeGreaterThanOrEqual(next.amount)
+      }
+    })
+  })
+
+  describe('setRichListLength()', () => {
+    it('should control rich list query output length', async () => {
+      richListCore.start()
+      await waitForCatchingUp(richListCore)
+
+      richListCore.setRichListLength(3)
+      await richListCore.calculateNext()
+
+      const richList = await richListCore.get('-1')
+      expect(richList.length).toStrictEqual(3)
+
+      for (let i = 0; i < richList.length - 1; i++) {
+        const current = richList[i]
+        const next = richList[i + 1]
+        expect(EXPECTED_RICH_LIST_ADDRESSES).toContain(current.address)
+        expect(EXPECTED_RICH_LIST_ADDRESSES).toContain(next.address)
+        expect(current.amount).toBeGreaterThanOrEqual(next.amount)
+      }
+
+      const excludedFromTopThree = await richListCore.addressBalances.list({
+        partition: '-1', // token id for utxo
+        order: 'DESC',
+        limit: Number.MAX_SAFE_INTEGER,
+        lt: new BigNumber(richList[2].amount).times('1e8').dividedToIntegerBy(1).toNumber()
+      })
+      expect(excludedFromTopThree.length).toStrictEqual(EXPECTED_RICH_LIST_ADDRESSES.length - 3)
+
+      for (let i = 0; i < excludedFromTopThree.length; i++) {
+        expect(excludedFromTopThree[i].data.amount).toBeLessThanOrEqual(richList[2].amount)
+      }
+    })
+  })
+})
```

Comment:
Probably can add a test for block invalidation + rich list updating logic? Or can do in a follow-up PR

I tried the following, but the balance doesn't get updated when fundAddress is called. I might be missing something / misunderstanding.
```suggestion
})


  describe('invalidation', () => {
    it('should update balances when block is invalidated', async () => {
      // Given some addresses having balances
      richListCore.start()
      await waitForCatchingUp(richListCore)
      await richListCore.calculateNext()

      // When one of the rich gets a lot of DFI
      await container.fundAddress('2NCWAKfEehP3qibkLKYQjXaWMK23k4EDMVS', 9_000_000)
      await container.generate(1)

      // And the address queue is processed
      await waitForCatchingUp(richListCore)
      await richListCore.calculateNext()
      const queue = await richListCore.queueClient.createQueueIfNotExist('RichListCore_ACTIVE_ADDRESSES', 'LIFO')
      const anythingRemainedInQ = await queue.receive(Number.MAX_SAFE_INTEGER)
      expect(anythingRemainedInQ.length).toStrictEqual(0)

      // Then rich list is updated
      const richList = await richListCore.get(UTXO)
      expect(richList.length).toStrictEqual(EXPECTED_RICH_LIST_ADDRESSES.length)
      expectCorrectRichListOrderedByBalance(richList)

      // When block is invalidated
      // ...
      // Then rich list is updated again
    })
  })
})

function expectCorrectRichListOrderedByBalance(richList: AddressBalance[]) {
  for (let i = 0; i < richList.length - 1; i++) {
    const current = richList[i]
    const next = richList[i + 1]
    expect(EXPECTED_RICH_LIST_ADDRESSES).toContain(current.address)
    expect(EXPECTED_RICH_LIST_ADDRESSES).toContain(next.address)
    expect(current.amount).toBeGreaterThanOrEqual(next.amount)
  }
}
```

Reply:
The logic makes sense, technically the `invalidate()` execution itself can be tested. But since it is a private func,  should be handled in some newer PR with e2e.

Reply:
One more thing to check in this test - address '2NCWAKfEehP3qibkLKYQjXaWMK23k4EDMVS' should be first in the list

Yea let's have a separate PR with more tests then :shipit:

Reply:
The tests is currently being added at #1224, along with tests for rich-list-api